server/api/src/rsDataObjClose.cpp

```diff
@@ -728,6 +729,31 @@ int rsDataObjClose(rsComm_t* rsComm, openedDataObjInp_t* dataObjCloseInp)
 
     int ec = 0;
 
+    // Call notify on resource plugin if this is a write. This allows S3 to pave over the physical path
+    // before it is written to the database in case decoupled mode is being used.  See irods/irods_resource_plugin_s3#2146.
+    if ((l1desc.dataObjInp->openFlags & O_ACCMODE) != O_RDONLY) {
+
+        irods::file_object_ptr file_obj(
+            new irods::file_object(
+                rsComm,
+                L1desc[fd].dataObjInfo));
+
+        irods::error ret = fileNotify(
+            rsComm,
+            file_obj,
+            irods::WRITE_OPERATION);
+
+        if ( !ret.ok() ) {
+            std::stringstream msg;
+            msg << "Failed to signal the resource that the data object \"";
+            msg << L1desc[fd].dataObjInfo->objPath;
+            msg << "\" was modified.";
```

Comment:
I know this came from DataObjWrite, but consider replacing this with `fmt::format`.